pkg/storage/badger/badger.go

```diff
@@ -421,6 +423,20 @@ func (st *Storage) scanIndexKeys(indexKey []byte, createdAtMin, createdAtMax tim
 	return keys, err
 }
 
+// cleanUpLoop removes expired values from the badger log. Code is taken from
+// https://github.com/dgraph-io/badger#garbage-collection
+func (st *Storage) cleanUpLoop() {
+	ticker := time.NewTicker(5 * time.Minute)
+	defer ticker.Stop()
+	for range ticker.C {
+	again:
+		err := st.db.RunValueLogGC(0.7)
+		if err == nil {
+			goto again
+		}
+	}
```

Comment:
I'm a bit puzzled with that change, it seems that it is the same as below, isn't it?

```go
for {
  err := st.db.RunValueLogGC(0.7)
  if err != nil {
    time.Sleep(5mins)
  }
}
```

Is that expected that in the happy path, where no errors returned, the goroutine will keep spinning in the busy loop?

I believe these errors should be logged with `logger.Error`, as such kind of errors should warn the operator. 

Also, what is the behaviour of the GC for the case when the db was closed, e.g. during application restart? Do we need to be careful and care to stop the cleanup goroutine, before we call `db.Close`?

Reply:
The GC is run until `ErrNoRewrite` is returned. In that case, the job is done and the goroutine can pause.
Since the storage interface `storage.Writer` and `storage.Reader` doesn't have any `Close` method, I expected, the storages will run forever and do not need a teardown.

Reply:
Yes, storage currently don't care about the closing procedure, but the storage accepts a pre-configured badger instance in its constructor. It's the responsibility of the caller, currently, `main` function, to handle the shutdown.

I'm thinking, how about instead of spawning the clenup goroutine inside the storage, we did that in [`initBadgerStorage` function](https://github.com/profefe/profefe/blob/master/cmd/profefe/main.go#L95). With that we would also be able to easily configure the GC percentage and cleanup intervals via config. What do you think?

Reply:
In the current architecture, that would be the best solution. I'll do the changes.

In general, I'd recommend to move all badger-related code into the `badger` package. Further I'd combine the `storage.{Writer,Reader}` into a single interface and add a `Close` method. By that all storage-specific code can be hidden behind that interface and the caller wouldn't have to care. But that might be part of a different PR.

Reply:
My biggest struggle with moving all the logic, that is specific to a DB, inside the `storage` interface, is that I don't see a clear general API, that can be wrapped in that interface. For example, 
- what would `Close` do for a storage implementation on top of S3-like blob storage or the raw FS;
- how testing of the storage would look like if the database was be buried inside the private fields (I will switch testing with badger's inmemory mode, later)

I hope to find a sweet spot, where interfaces are small enough to be flexible enough, moving all the specific details as far as possible. But this is likely very debatable.